pandas/core/indexers.py

```diff
@@ -307,3 +312,62 @@ def check_bool_array_indexer(array: AnyArrayLike, mask: AnyArrayLike) -> np.ndar
     if len(result) != len(array):
         raise IndexError(f"Item wrong length {len(result)} instead of {len(array)}.")
     return result
+
+
+def check_array_indexer(array, indexer) -> np.ndarray:
+    """
+    Check if `indexer` is a valid array indexer for `array`.
+
+    `array` and `indexer` are checked to have the same length, and the
+    dtype is validated. If it is an integer or boolean ExtensionArray, it is
+    checked if there are missing values present, and it is converted to
+    the appropriate numpy array.
+
+    .. versionadded:: 1.0.0
+
+    Parameters
+    ----------
+    array : array
+        The array that's being indexed (only used for the length).
+    indexer : array-like
```

Comment:
In a few places above, you've done `is_list_like`, but here we require an array (with a dtype).

Thoughts on what we want? Requiring an array is certainly easier, so that we don't have to infer the types. But users may be passing arbitrary objects to `__getitem__`.

Reply:
We actually don't require an array with a dtype. The first thing that this function does is:

```
    if not is_array_like(indexer):
        indexer = pd.array(indexer)
```

to deal with eg lists.

So I probably meant to update the array into "list-like" instead of "array-like"